examples/regression/2-advanced-analysis/plot_conditional_coverage.py

```diff
@@ -0,0 +1,319 @@
+# coverage. If we interpret the result for the four methods here : CV+ seems to
+# be the better one. And with the hsic correlation coefficient, we have the
+# same interpretation : :func:`~mapie.metrics.hsic` computes the correlation
+# between the coverage indicator and the interval size, a value of 0
+# translates an independence between the two.
+#
+# We would like to highlight here the misinterpretation that can be made
+# with these metrics. In fact, here CV+ with the absolute residual score
+# calculates constant intervals which, by definition, are not adaptive.
+# Therefore, it is very important to check that the intervals widths are well
+# spread before drawing conclusions (with a plot of the distribution of
+# interval widths or a visualisation of the data for example).
+
+
+# Visualition of the data and predictions
+def plot_intervals(X, y, y_pred, intervals, title="", ax=None):
+    """
+    Plots the data X, y with associated intervals and predictions points.
+
+    Parameters
+    ----------
+    X: ArrayLike
+        Observed features
+    y: ArrayLike
+        Observed targets
+    y_pred: ArrayLike
+        Predictions
+    intervals: ArrayLike
+        Prediction intervals
+    title: str
+        Title of the plot
+    ax: matplotlib axes
+        An ax can be provided to include this plot in a subplot
+    """
+    if ax is None:
+        fig, ax = plt.subplots(figsize=(6, 5))
+
+    y_pred = y_pred.reshape((X.shape[0], 1))
+    order = np.argsort(X[:, 0])
+
+    # data
+    ax.scatter(X.ravel(), y, color="#1f77b4", alpha=0.3, label="data")
+    # predictions
+    ax.scatter(
+        X.ravel(), y_pred,
+        color="#ff7f0e", marker="+", label="predictions", alpha=0.5
+    )
+    # intervals
+    for i in range(intervals.shape[-1]):
+        ax.fill_between(
+            X[order].ravel(),
+            intervals[:, 0, i][order],
+            intervals[:, 1, i][order],
+            color="#ff7f0e",
+            alpha=0.3
+        )
+
+    ax.set_xlabel("x")
+    ax.set_ylabel("y")
+    ax.set_title(title)
+    ax.legend()
+
+
+def plot_coverage_by_width(y, intervals, num_bins, title="", ax=None):
+    """
+    PLots a bar diagram of coverages by groups of interval widths.
+
+    Parameters
+    ----------
+    y: ArrayLike
+        Observed targets.
+    intervals: ArrayLike
+        Intervals of prediction
+    num_bins: int
+        Number of groups of interval widths
+    title: str
+        Title of the plot
+    ax: matplotlib axes
+        An ax can be provided to include this plot in a subplot
+    """
+    if ax is None:
+        fig, ax = plt.subplots(figsize=(6, 5))
+
+    ax.bar(
+        np.arange(num_bins),
+        regression_ssc(y, intervals, num_bins=num_bins)[0]
+    )
+    ax.set_title(title)
+    ax.set_xlabel("intervals grouped by size")
+    ax.set_ylabel("coverage")
+    ax.tick_params(
+        axis='x', which='both', bottom=False, top=False, labelbottom=False
+    )
```

Comment:
SUGGESTION: could you add a line representing the target coverage (0.9)? That would be a very relevant visual clue, in line with your analysis.

Reply:
Done